pkg/backoff/backoff.go

```diff
@@ -0,0 +1,123 @@
+package backoff
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	ini "github.com/wakatime/wakatime-cli/pkg/config"
+	"github.com/wakatime/wakatime-cli/pkg/heartbeat"
+	"github.com/wakatime/wakatime-cli/pkg/log"
+
+	"github.com/spf13/viper"
+)
+
+const (
+	// resetAfter sets the total seconds a backoff will last.
+	resetAfter = 3600
+	// factor is the total seconds to be multiplied by.
+	factor = 15
+)
+
+// RetryConfig defines backoff data.
+type RetryConfig struct {
+	// At is the time when the first failure happened.
+	At *time.Time
+	// Retries is the number of attempts to connect.
+	Retries int
+	// V is an instance of Viper.
+	V *viper.Viper
+}
+
+// WithRetry initializes and returns a heartbeat handle option, which
```

Comment:
We prevent sending heartbeats here, so actually we're implementing a circuit breaker here (https://microservices.io/patterns/reliability/circuit-breaker.html).

It is not really about a retrying here. Retrying only happens due to offline pipeline step. But the backoff pipeline step should not depend on offline step being present or not. So how about changing the name to `WithBackoff` or something  similar? 

```suggestion
// WithBackoff initializes and returns a heartbeat handle option, which
```

Reply:
humm not to sure if it's a circuit breaker. Most of the articles refer backoff as a retry strategy and circuit breaker as a gate to return success or not in case of failures. Even if wakatime-cli runs once per request (instead of an API that is always on) it controls backoff by applying a retry policy persisted as `backoff_retries` and backoff_at`. IMHO it's more a retry than a circuit breaker. Think about offline step, it also breaks the execution if fails.

Reply:
But this particular step does not control the retry. The only thing it does is to block the call or not and increment a count.

Reply:
renamed